hudi-io/src/main/java/org/apache/hudi/common/util/StringUtils.java

```diff
@@ -369,6 +370,53 @@ public static String concatenateWithThreshold(String a, String b, int byteLength
     return a.substring(0, bestLength) + b;
   }
 
+  /**
+   * Concatenates the string representation of each object in the list
+   * and returns a single string. The result is capped at "lengthThreshold" characters.
+   *
+   * @param objectList      object list
+   * @param lengthThreshold number of characters to cap the string
+   * @param <T>             type of the object
+   *
+   * @return capped string representation of the object list
+   */
+  public static <T> String toStringWithThreshold(List<T> objectList, int lengthThreshold) {
+    if (objectList == null || objectList.isEmpty()) {
+      return "";
```

Comment:
another corner case: what if `lengthThreshold` is less than 3? IIUC, the method will attempt to set the last three characters of the string as dots, which could result in unexpected behavior. Maybe we just call the usual toString when length is under 300 characters? No need to go through this logic for short messages?

Reply:
`StringUtils#setLastThreeDots` method has a safeguard to make it does not trigger index out-of-bound.

Yes, this method is only intended to be called with a potential large list.  In normal cases this method is not necessary.  The logic does not directly call `objectList.toStirng()` because that can go through every item in the list and generate a huge String which is not necessary.  The current logic stops at the item in the list that fills the characters to the `StringBuilder` up to `lengthThreshold`.